firefox_puppeteer/api/software_update.py

```diff
@@ -0,0 +1,491 @@
+
+        :returns: JSON data for the selected patch
+        """
+
+        return self.marionette.execute_script("""
+          let ums = Cc['@mozilla.org/updates/update-manager;1']
+                    .getService(Ci.nsIUpdateManager);
+          return ums.activeUpdate.selectedPatch;
+        """)
+
+
+class MARChannels(BaseLib):
+    """Class to handle the allowed MAR channels as listed in update-settings.ini
+    """
+
+    INI_SECTION = 'Settings'
+    INI_OPTION = 'ACCEPTED_MAR_CHANNEL_IDS'
+
+    def __init__(self, marionette_getter):
+        BaseLib.__init__(self, marionette_getter)
+
+        self.ini_file_path = self.marionette.execute_script("""
+          Cu.import('resource://gre/modules/Services.jsm');
+
+          let file = Services.dirsvc.get('GreD', Ci.nsIFile);
+          file.append('update-settings.ini');
+
+          return file.path;
+          """)
+
+    @property
+    def config_file_path(self):
+        return self.ini_file_path
+
+    @property
+    def config_file_contents(self):
+        with open(self.ini_file_path) as f:
+            return f.read()
+
+    @property
+    def channels(self):
+
+        channels = self.marionette.execute_script("""
+          Cu.import('resource://gre/modules/Services.jsm');
+
+          let iniFactory = Cc['@mozilla.org/xpcom/ini-processor-factory;1']
+                           .getService(Ci.nsIINIParserFactory);
+
+          let file = Services.dirsvc.get('GreD', Ci.nsIFile);
+          file.append('update-settings.ini');
+          let parser = iniFactory.createINIParser(file);
+
+          return parser.getString(arguments[0], arguments[1]);
+          """, script_args=[self.INI_SECTION, self.INI_OPTION])
+        return channels.split(',')
+
+    @channels.setter
+    def channels(self, channels):
+        """Set the channels in the ini file
+
+        :param channels: a comma delimited string of channel names
+        """
+
+        self.marionette.execute_script("""
+          Cu.import('resource://gre/modules/Services.jsm');
+
+          let iniFactory = Cc['@mozilla.org/xpcom/ini-processor-factory;1']
+                           .getService(Ci.nsIINIParserFactory);
+
+          let file = Services.dirsvc.get('GreD', Ci.nsIFile);
+          file.append('update-settings.ini');
+
+          let writer = iniFactory.createINIParser(file)
+                       .QueryInterface(Ci.nsIINIParserWriter);
+
+          writer.setString(arguments[0], arguments[1], arguments[2]);
+          writer.writeFile(null, Ci.nsIINIParserWriter.WRITE_UTF16);
+          """, script_args=[self.INI_SECTION, self.INI_OPTION, channels])
+
+    def add_channels(self, channels):
+        new_channels = ','.join(self.channels) + ',' + channels
+        self.channels = new_channels
+
+    def remove_channels(self, channels):
+        current_channels = self.channels
+        new_channels = [channel for channel in current_channels if channel not in channels]
```

Comment:
When looking at this it reminds me of a comment from @chmanchester and that we should better use `Set() - Set()` here. Maybe we can hold `self.channels` as a set on that class directly.